### PR TITLE
Added test coverage for Symbol class

### PR DESCRIPTION
**Pull request by JckXia**
Added new tests for the Symbol class. Also added a static method to the Symbol class to perform "Symbol.For()" from the C++ side, since the exisiting method (Symbol::Wellknown) , which has a different functionality caused some confusion.

### CODE REVIEW

**Comment by JckXia** [comment]
I can't seem to recreate this test failure. It seems intermittent as the CI running against this commit is successful in my forked repo. https://github.com/JckXia/node-addon-api/actions/runs/765661855.

File: napi.h
Comment (NickNaso) [inline napi.h:542]: Do we need some other overloads for this methods?
```cpp
static Symbol For(napi_env env,  const char* description = nullptr);
static Symbol For(napi_env env,  String description);
static Symbol For(napi_env env, napi_value description);
```
What do you think about?

Comment (NickNaso) [inline napi.h:542]: Please remember to add documentation for this new method.

File: test/symbol.cc
Comment (NickNaso) [inline test/symbol.cc:6]: Why does it necessary do this?
```cpp
(void)info;
```

File: test/symbol.js
Comment (NickNaso) [inline test/symbol.js:65]: Please remove one empty new line.

File: test/symbol.cc
Comment (NickNaso) [inline test/symbol.cc:51]: PLease add one empty new line.

**Review by NickNaso — COMMENTED** [review]
@JckXia well done. Just some comments.

**Comment by NickNaso** [comment]
> I can't seem to recreate this test failure. It seems intermittent as the CI running against this commit is successful in my forked repo. https://github.com/JckXia/node-addon-api/actions/runs/765661855.

Don't worry I restarted the jobs and now it's all green.

File: test/symbol.cc
Comment (JckXia) [reply test/symbol.cc:6]: I think this was to suppress compiler warnings for unused parameter, since the default Symbol constructor doesn't take in any values. But i can remove this function since it's not being called from js.

File: napi.h
Comment (JckXia) [reply napi.h:542]: Hmm yeah I think that will be a good idea. Will add them in soon.

File: test/symbol.cc
Comment (NickNaso) [reply test/symbol.cc:6]: To avoid compiler complains maybe you can try the following solution (comment the formal parameter of the function):
```cpp
Symbol CreateNewSymbolWithNoArgs(const Napi::CallbackInfo& /*info*/) {
```
Let me know if it worked.

Comment (JckXia) [reply test/symbol.cc]: Yep it works. Thanks!

File: doc/symbol.md
Comment (NickNaso) [inline doc/symbol.md:37]: Please fix the typo it should be **WellKnown**

File: doc/symbol.md
Comment (NickNaso) [inline doc/symbol.md:52]: Please add the documentation for all the overloads of `Napi::Symbol Napi::Symbol::WellKnown()` methos.

File: doc/symbol.md
Comment (NickNaso) [inline doc/symbol.md:56]: Maybe the following description is better: 
Searches in the global registry for existing symbol with the given name. If the symbol already exist it will be returned, otherwise a new symbol will be created in the registry. It's equivalent to `Symbol.for()` called from JavaScript.

File: napi-inl.h
Comment (NickNaso) [inline napi-inl.h:1014]: You can directory pass `description` to `Symbol::For()` because `Napi::String` is a `Napi::Value` that has the cast operator `napi_value`.

**Review by NickNaso — COMMENTED** [review]
Hi @JckXia,
I added som e comments about you last work.

File: doc/symbol.md
Comment (KevinEady) [inline doc/symbol.md:50]: You are adding the `For` method... should this be `For`...?

File: doc/symbol.md
Comment (NickNaso) [inline doc/symbol.md:37]: It shlould be `WellKnown` insteand of `Wellknown`. The `k` should be in uppercase.

File: doc/symbol.md
Comment (NickNaso) [inline doc/symbol.md:53]: Please use the complete namespace there:

```cpp
static Napi::Symbol Napi::Symbol::For(napi_env env, const char* description = nullptr);
static Napi::Symbol Napi::Symbol::For(napi_env env, String description);
static Napi::Symbol Napi::Symbol::For(napi_env env,napi_value description);

```

**Review by NickNaso — COMMENTED** [review]
Hi @JckXia,
just something to fix.

**Review by NickNaso — COMMENTED** [review]
Hi @JckXia,
could you propose the last change on the package.json on a different PR? I'm asking for this because the change it's not correlated with the test and in case we will need to revert the change it will be more simpler to do. For the rest it's all ok.

**Comment by JckXia** [comment]
Hey @NickNaso 
Thanks for the review! I think I accidentally made some changes to package.json and checked it into git for some reason and was trying to revert it.  There was no changes needed to be made to package.json and I will try to revert the changes made to it. Sorry about that.

**Review by NickNaso — APPROVED** [review]
LGTM

File: napi.h
Comment (gabrielschulhof) [inline napi.h:552]:
```suggestion
   static Symbol For(Env env, const std::string& description);

   // Create a symbol in the global registry, C style string (null terminated)
   static Symbol For(Env env, const char* description = nullptr);

   // Create a symbol in the global registry, String value describing the symbol
   static Symbol For(Env env, String description);

   // Create a symbol in the global registry, napi_value describing the symbol
   static Symbol For(Env env, napi_value description);

```

File: napi-inl.h
Comment (gabrielschulhof) [inline napi-inl.h:1001]:
```suggestion
inline Symbol Symbol::For(Napi::Env env, const std::string& description) {
```

... and a similar change for all APIs below.

**Review by gabrielschulhof — CHANGES_REQUESTED** [review]
Blocking until we figure out what happens when you pass `nullptr` to `napi_create_string_utf8()` as its `const char*` parameter.

**Comment by gabrielschulhof** [comment]
@JckXia thanks for catching this!

**Comment by gabrielschulhof** [comment]
We must never pass `nullptr` into `napi_create_string_*()` because that will cause a crash. Thus, we should actually change *all* the APIs (even existing ones) that end up calling `napi_create_string_*()` to return `napi_invalid_arg` when passed in what evaluates to `(const char *)nullptr`.

I'll make a PR upstream to fix this in core.

**Comment by JckXia** [comment]
Thanks for the review @gabrielschulhof . Just making sure, are we converting all function signature from For(Napi::Env env, const std::string & val),etc to For(napi_env env, const std::string & val), or are we keeping both?

**Comment by gabrielschulhof** [comment]
I checked napi.h by looking for `[(]napi_env` and it turns out most of our APIs take `napi_env` as the first argument so it's OK to keep the `napi_env` and we need neither replace nor provide overloads. So, LGTM except for the passing-null-to-`napi_create_string_*` part.

**Comment by gabrielschulhof** [comment]
Refs: https://github.com/nodejs/node/pull/38923

**Comment by mhdawson** [comment]
@JckXia just wondering what the next steps are to get this landed?

**Comment by JckXia** [comment]
@mhdawson  I believe that it is ready to be merged.

File: doc/symbol.md
Comment (KevinEady) [inline doc/symbol.md:51]: Hi @JckXia ,

Do we have a test for this `nullptr` value...? I am not sure what happens when using a nullptr `napi_value` for the `description` here:

```cpp
  auto forSymb =
      symbObject.Get("for").As<Function>().Call(symbObject, {description});
```

and I think a test would be beneficial

**Comment by JckXia** [comment]
@KevinEady Thanks for catching that! Apparently if we simply pass the C++ nullptr into that function it will cause a SegFault. I updated the Symbol::For method impl to use the napi_value we get from calling the napi_get_null function instead and it seems to work.

**Comment by KevinEady** [comment]
> @KevinEady Thanks for catching that! Apparently if we simply pass the C++ nullptr into that function it will cause a SegFault. I updated the Symbol::For method impl to use the napi_value we get from calling the napi_get_null function instead and it seems to work.

@JckXia  Aahahh... I am inclined then to remove the `std::nullptr` default and the "check if nullptr, assume napi_get_null" ... Because that is not something we have in other locations. For example,

```js
{
  [null]: "3"
}
```
is a valid JSON object, but we do not check in `Object::Get` for using `const char* utf8name` with `std::nullptr`:

https://github.com/nodejs/node-addon-api/blob/ad76ad07f914fab02be5778ec67485916c4626d9/napi-inl.h#L1227-L1232

The application would just crash (I assume). I would think, for consistency purposes, we should remove the nullptr default as well as the runtime check (which is additional overhead anyway).

If the user wanted to explicitly use `null`, they could use `Symbol::For(env, env.Null())`.

Thoughts?

**Comment by JckXia** [comment]
@KevinEady Oh okay I see. So I think we can simply pass the description argument into String::New(), and pass the newly constructed Napi string the Symbol::For function?  Also, I think this ` Symbol::For(env, env.Null())` is a good approach, as I believe they could do something similar for `undefined`. Will update the PR and later today. :)

**Comment by KevinEady** [comment]
This depends on landing #1015 first.

**Comment by gabrielschulhof** [comment]
Please feel free to dismiss my change request after #1015 has landed and the `nullptr` default for the `char *` argument was removed!

**Comment by KevinEady** [comment]
As discussed in today's meeting, please remove the `std::nullptr` default value in `Symbol::For`.

**Comment by mhdawson** [comment]
1015 landed, dismissing @gabrielschulhof review as suggested above.

**Review by mhdawson — APPROVED** [review]
LGTM

**Comment by JckXia** [comment]
@mhdawson Thanks, it's been fixed.

**Comment by mhdawson** [comment]
Landed in e02e8a4ce3c24f9e833b771cff7aac51658cf2ee

**Pull request closed without merging**
